mlflow/_spark_autologging.py

```diff
@@ -89,57 +92,67 @@ def _set_run_tag(run_id, path, version, data_format):
     client.set_tag(run_id, _SPARK_TABLE_INFO_TAG_NAME, new_table_info)
 
 
+def _listen_for_spark_activity(spark_context):
+    if _get_spark_major_version(spark_context) < 3:
+        raise MlflowException("Spark autologging unsupported for Spark versions < 3")
+
+    gw = spark_context._gateway
+    params = gw.callback_server_parameters
+    callback_server_params = CallbackServerParameters(
+        address=params.address,
+        port=params.port,
+        daemonize=True,
+        daemonize_connections=True,
+        eager_load=params.eager_load,
+        ssl_context=params.ssl_context,
+        accept_timeout=params.accept_timeout,
+        read_timeout=params.read_timeout,
+        auth_token=params.auth_token,
+    )
+    gw.start_callback_server(callback_server_params)
+
+    event_publisher = _get_jvm_event_publisher()
+    try:
+        event_publisher.init(1)
+        _spark_table_info_listener = PythonSubscriber()
+        _spark_table_info_listener.register()
+    except Exception as e:
+        gw.shutdown_callback_server()
+        raise MlflowException(
+            "Exception while attempting to initialize JVM-side state for "
+            "Spark datasource autologging. Please create a new Spark session "
+            "and ensure you have the mlflow-spark JAR attached to your Spark "
+            "session as described in "
+            "http://mlflow.org/docs/latest/tracking.html#automatic-logging-from-spark-experimental. "
+            "Exception:\n%s" % e
+        )
+
+    # Register context provider for Spark autologging
+    from mlflow.tracking.context.registry import _run_context_provider_registry
+
+    _run_context_provider_registry.register(SparkAutologgingContext)
```

Comment:
in the case where `_get_current_listener() is None`, we should not trigger this `_run_context_provider_registry.register(SparkAutologgingContext)` according to the existing logic. Leaving this code inside `_listen_for_spark_activity` allows us to maintain this logic, otherwise if we move it outside we would need to check for `get_current_listener()` a second time